src/components/application_manager/src/request_tracker.cc

```diff
@@ -0,0 +1,127 @@
+* modification, are permitted provided that the following conditions are met:
+*
+* Redistributions of source code must retain the above copyright notice, this
+* list of conditions and the following disclaimer.
+*
+* Redistributions in binary form must reproduce the above copyright notice,
+* this list of conditions and the following
+* disclaimer in the documentation and/or other materials provided with the
+* distribution.
+*
+* Neither the name of the Ford Motor Company nor the names of its contributors
+* may be used to endorse or promote products derived from this software
+* without specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+* ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+* LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+* POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include "utils/logger.h"
+#include "utils/macro.h"
+#include "application_manager/request_tracker.h"
+
+namespace application_manager {
+
+namespace request_controller {
+
+CREATE_LOGGERPTR_GLOBAL(logger_, "RequestController")
+
+RequestTracker::RequestTracker(const RequestControlerSettings& settings)
+    : settings_(settings) {}
+
+TrackResult RequestTracker::Track(const ApplicationID& app_id,
+                                  const mobile_apis::HMILevel::eType level) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  bool track_result = false;
+
+  LOG4CXX_DEBUG(logger_, "Tracking request for level: " << level);
+
+  if (mobile_apis::HMILevel::HMI_NONE == level) {
+    track_result = Track(app_id,
+                         settings_.app_hmi_level_none_time_scale(),
+                         settings_.app_hmi_level_none_time_scale_max_requests(),
+                         none_level_tracker_);
+
+    return track_result ? kSuccess : kNoneLevelMaxRequestsExceeded;
+  }
+
+  track_result = Track(app_id,
+                       settings_.app_time_scale(),
+                       settings_.app_time_scale_max_requests(),
+                       tracker_);
+
+  return track_result ? kSuccess : kMaxRequestsExceeded;
+}
+
+bool RequestTracker::Track(const ApplicationID& app_id,
+                           const uint32_t time_scale,
+                           const uint32_t max_requests,
+                           ApplicationsRequestsTracker& tracker) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  if (!time_scale || !max_requests) {
```

Comment:
@dev-gh I completely disagree with you in that point. 
```time_scale && max_requests ``` requires much less time for analyses